signature/src/balrog.rs

```diff
@@ -0,0 +1,382 @@
+ * can be trusted, and verifies a content signature against the code signing
+ * certificate.
+ *
+ * The content signature itself is a base64-encoded ECDSA signature, using the
+ * content prefixed with "Content-Signature\x00" as the input data.
+ */
+
+use hex;
+use asn1_rs::ToDer;
+use ring::digest;
+use x509_parser::prelude::*;
+
+use oid_registry::{
+    OID_SIG_ECDSA_WITH_SHA256,
+    OID_SIG_ECDSA_WITH_SHA384,
+};
+
+pub struct Balrog<'a> { 
+    pub chain: Vec<X509Certificate<'a>>,
+    pub root_hash: Vec<u8>
+}
+
+// Errors that can be returned from the Balrog module.
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum BalrogError {
+    #[error("Chain subject/issuer mismatch")]
+    ChainSubjectMismatch,
+    #[error("Certificate expired")]
+    CertificateExpired,
+    #[error("Issuer unauthorized for certificate signing")]
+    IssuerUnauthorized,
+    #[error("Certificate unauthorized for code signing")]
+    CodeSignUnauthorized,
+    #[error("Root hash parse failed")]
+    RootHashParseFailed,
+    #[error("Root hash mismatch")]
+    RootHashMismatch,
+    #[error("Hostname mismatch")]
+    HostnameMismatch,
+    #[error("Signature decoding failed")]
+    SignatureDecodeError,
+
+    #[error("PEM decoding error: {0:?}")]
+    PemDecodeError(String),
+    #[error("X509 validation failed: {0:?}")]
+    X509ValidationError(String),
+    #[error("X509 error: {0:?}")]
+    X509(X509Error)
+}
+
+impl From<X509Error> for BalrogError {
+    fn from(e: X509Error) -> Self {
+        BalrogError::X509(e)
+    }
+}
+
+impl From<PEMError> for BalrogError {
+    fn from(e: PEMError) -> Self {
+        BalrogError::PemDecodeError(e.to_string())
+    }
+}
+
+pub fn parse_pem_chain(input: &[u8]) -> Result<Vec<Pem>, PEMError> {
+    Pem::iter_from_buffer(input)
+        .collect::<Result<Vec<Pem>, PEMError>>()
+}
+
+/* A one-and-done function that parses a signature chain and validates a content signature. */
+pub fn parse_and_verify(
+    x5u: &[u8],
+    input: &[u8],
+    signature: &str,
+    current_time: i64,
+    root_hash: &str,
+    leaf_subject: &str
+) -> Result<(), BalrogError> {
+    let pem_chain = match parse_pem_chain(x5u) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    let balrog = match Balrog::new(&pem_chain) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    let _ = match balrog.verify_chain(current_time, root_hash) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    let _ = match balrog.verify_leaf_hostname(leaf_subject) {
+        Err(e) => return Err(e),
+        Ok(x) => x
+    };
+
+    let _ = match balrog.verify_content_signature(input, signature) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    Ok(())
+}
+
+impl<'a> Balrog<'_> {
+    pub fn new(list: &'a Vec<Pem>) -> Result<Balrog<'a>, X509Error> {
```

Comment:
I think we can do without the lifetime annotations here. Why were they included?

Reply:
Would it make more sense for this return a BalrogError instead? The conversion would happen implicitly anyways.

Reply:
The lifetime annotations seem to be required for the `X509Certificate` type. I kind of just added them to make the compiler happy without really understanding what they do :)

Reply:
I always feel like lifetime annotations make Rust look so scary, also it's weird this would be required given `Vec`s own their data and store them in the heap. If removing it causes errors 🤷‍♀️  No real harm done other than me being scared lol.